L1Trigger/L1THGCal/src/be_algorithms/HGCalShowerShape.cc

```diff
@@ -0,0 +1,554 @@
+#include "TMath.h"
+#include <cmath>
+#include "DataFormats/L1THGCal/interface/HGCalMulticluster.h"
+#include "DataFormats/L1THGCal/interface/HGCalCluster.h"
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+
+#include <iostream>
+#include <sstream>
+#include <vector>
+#include <string>
+
+
+
+const float HGCalShowerShape::sigmaEtaEta(const std::vector<float> energy, const std::vector<float> eta){
+
+    int ntc=energy.size();
+    //compute weighted eta mean
+    float eta_sum=0;
+    float w_sum=0; //here weight is energy
+    for(int i=0;i<ntc;i++){
+        eta_sum+=energy.at(i)*eta.at(i);
+        w_sum+=energy.at(i);
+    }
+    float eta_mean=eta_sum/w_sum;
+    //compute weighted eta RMS
+    float deltaeta2_sum=0;
+    for(int i=0;i<ntc;i++) deltaeta2_sum+=energy.at(i)*pow((eta.at(i)-eta_mean),2);
+    float eta_RMS=0;
+    if (w_sum!=0) eta_RMS=deltaeta2_sum/w_sum;
+    float See=sqrt(eta_RMS);
+    return See;
+    
+}
+
+
+const float HGCalShowerShape::sigmaPhiPhi(const std::vector<float> energy, const std::vector<float> phi){
+
+    int ntc=energy.size();
+    //compute weighted phi mean
+    float phi_sum=0;
+    float w_sum=0; //here weight is energy
+    for(int i=0;i<ntc;i++){
+        if(phi.at(i)>0) phi_sum+=energy.at(i)*phi.at(i);
+        else phi_sum+=energy.at(i)*(phi.at(i)+2*TMath::Pi());
+        w_sum+=energy.at(i);
+    }
+    float phi_mean=phi_sum/w_sum;
+    if(phi_mean>TMath::Pi()) phi_mean-=2*TMath::Pi();
+    //compute weighted eta RMS
+    float deltaphi2_sum=0;
+    for(int i=0;i<ntc;i++){
+        float deltaPhi=std::abs(phi.at(i)-phi_mean);
+        if (deltaPhi>TMath::Pi()) deltaPhi=2*TMath::Pi()-deltaPhi;
+        deltaphi2_sum+=energy.at(i)*pow(deltaPhi,2);
+    }        
+    float phi_RMS=0;
+    if (w_sum!=0) phi_RMS=deltaphi2_sum/w_sum;
+    float Spp=sqrt(phi_RMS);
+    return Spp;
+    
+}
+
+
+
+const float HGCalShowerShape::sigmaZZ(const std::vector<float> energy, const std::vector<float> z){
+
+    int ntc=energy.size();
+    //compute weighted eta mean
+    float z_sum=0;
+    float w_sum=0; //here weight is energy
+    for(int i=0;i<ntc;i++){
+        z_sum+=energy.at(i)*z.at(i);
+        w_sum+=energy.at(i);
+    }
+    float z_mean=z_sum/w_sum;
+    //compute weighted eta RMS
+    float deltaz2_sum=0;
+    for(int i=0;i<ntc;i++) deltaz2_sum+=energy.at(i)*pow((z.at(i)-z_mean),2);
+    float z_RMS=0;
+    if (w_sum!=0) z_RMS=deltaz2_sum/w_sum;
+    float Szz=sqrt(z_RMS);
+    return Szz;
+    
+}
+
+
+const int HGCalShowerShape::showerLength(const l1t::HGCalMulticluster& c3d){
+
+    const edm::PtrVector<l1t::HGCalCluster> clustersPtrs = c3d.constituents();
+
+    int firstLayer=999;
+    int lastLayer=-999;
+    int layer=999;
+
+    int nLayersEE=28;
+   
+    for(const auto& clu : clustersPtrs){
+
+        if(clu->subdetId()==HGCEE) layer=clu->layer();//EE
+        if(clu->subdetId()==HGCHEF) layer=clu->layer()+nLayersEE;//FH
+
```

Comment:
* Better to use `switch` `case` compared to several consecutive `if`
* The `BH` should be added as well, otherwise the function will return things like `999` as soon as the cluster includes `BH` layers